snuba/datasets/storage.py

```diff
@@ -0,0 +1,104 @@
+from abc import ABC
+from typing import Optional, Sequence
+
+
+from snuba.datasets.dataset_schemas import StorageSchemas
+from snuba.datasets.table_storage import TableWriter
+from snuba.query.query import Query
+from snuba.query.query_processor import QueryProcessor
+from snuba.request.request_settings import RequestSettings
+
+
+class Storage(ABC):
+    """
+    Storage is an abstraction on anything we can run a query onto in our
+    database. This means that it generally represents a Clickhouse table
+    or a view.
+    It provides:
+    - what we need to build the query (the schemas)
+    - it can provide a table writer if we can write on this storage
+    - a sequence of query processors that are applied to the query after
+      the storage is selected.
+    There are one or multiple storages per dataset (in the future, there
+    will be multiple per entity). During the query processing a storage
+    is selected and the query focuses on that storage from that point.
+    """
+
+    # TODO: Break StorageSchemas apart. It contains a distinction between write schema and
+    # read schema that existed before this dataset model and before TableWriters (then we
+    # trusted StorageSchemas to define which schema we would write on and which one we would
+    # read from). This is not needed anymore since TableWriter has its own write schema.
+    def get_schemas(self) -> StorageSchemas:
+        """
+        Returns the collections of schemas for DDL operations and for query.
+        See TableWriter to get a write schema.
+        """
+        raise NotImplementedError
+
+    def can_write(self) -> bool:
+        """
+        Returns True if this dataset has write capabilities
+        """
+        return self.get_table_writer() is not None
+
+    def get_table_writer(self) -> Optional[TableWriter]:
+        """
+        Returns the TableWriter if the Storage has one.
+        """
+        raise NotImplementedError
```

Comment:
yes that's in plan.

Reply:
After reading through more where the `Storage` instances are built, I think that the storage and loader/writer relationship seems backwards (the writer/loader should write _to_ the storage rather than being exposed _by_ the storage.) I don't think we need to do that right now (and definitely not in the scope of this specific change) but curious what your thoughts are around this.

Reply:
That would mean have another first level concept (the writer) in the dataset (later in the entity). Conceptually I am fine with that, from the implementation standpoint I cannot answer on that now, the writer provides the message processor and the replacer as well.
I would say replacer and message processor should be provided by the same entity, whether this is the writer or the storage, and since they are specific to the schema (ok, they are supposed to depend on the schema) I would say they should come from the storage and not from the Writer but I would not have a very strong opinion on this.
This would mean
Writer -> Storage -> processor/replacer

Does it make sense to you ?
If yes, then it is definitively a separate PR because it impacts consumers and replacers as well.

Reply:
> This would mean
> Writer -> Storage -> processor/replacer
> Does it make sense to you ?

That is pretty much my thinking, with the exception that I think the replacer could (and possibly should) be coupled more closely with storages that are backed by the replacing merge tree than the entity definition itself, at least for the easily generalizable case of deletions. We would probably need to be able to extend it to handle other types of replacements that are entity specific (merging and unmerging), column mutations like tag deletions probably fit somewhere in between.

I don't have a specific design for this and I don't think we need to rush into this (I think some of this will be fleshed out with the entities work as you mentioned) but I wanted to make sure that even if we're not exactly on the same page we're at least reading the same book so to speak and that there weren't other parts of the design that are going to strictly rely on this design as it exists today that I am not seeing.